### PR TITLE
[Fix] `jsx-sort-props`: sorted attributes now respect comments

### PR DESCRIPTION
**Pull request by ROSSROSALES**
Fixes https://github.com/jsx-eslint/eslint-plugin-react/issues/2366

Previously comments were not sorted with rule.
With fix, comments are included during sort.

Special Attribute Conditions:
- if comment found between two attributes, attributes and comment are sorted to end (to be manually sorted if necessary)
- above condition also applies to block comments
- single-line comments on same attribute-line are sorted

### CODE REVIEW

**Comment by ljharb** [comment]
Looks like there's older eslint versions where `sourceCode.getCommentsAfter` doesn't exist, so we'll need to handle that both in the rule, and in the test cases

**Review by ROSSROSALES — COMMENTED** [review]
Thank you for the review!
I really appreciate your time. I have learned many of my own faults with redundant code, naming variables, and conditional statements. I hope in the future to improve and reduce these nuances.

**Comment by ROSSROSALES** [comment]
> Looks like there's older eslint versions where `sourceCode.getCommentsAfter` doesn't exist, so we'll need to handle that both in the rule, and in the test cases

For this change:
Would we need to modify the code to identify which ESLint version is being used?
Then create another function using older deprecated methods of ESLint?

I am unsure other ways of handling errors for older ESLint versions.

**Comment by ROSSROSALES** [comment]
@ljharb 
I decided to go with latter of "work with newer ESLint versions". Other useful methods have been deprecated as well and would throw same previous error.

After changes,
I am getting a new error:
```AssertionError: Did not specify errors for an invalid test of jsx-sort-props```

I tried to search issue online and through other rule using older ESLint, but I had no luck. 
How is this error handled?

**Comment by ljharb** [comment]
The `invalid` list needs to be changed from an array, to a `[].concat()` call.

I'm thinking about whether or not we need an option for this, or whether it should be the default behavior. Either way, there aren't any tests for the `commentbetween` option - so, let's either add those, or remove the code handling that option?

**Comment by ROSSROSALES** [comment]
> The `invalid` list needs to be changed from an array, to a `[].concat()` call.
> 
> I'm thinking about whether or not we need an option for this, or whether it should be the default behavior. Either way, there aren't any tests for the `commentbetween` option - so, let's either add those, or remove the code handling that option?

I created the `commentbetween` as an option defaulted to true because of how I initially understood `sort((a, b) => contextCompare(a, b, options)))` . I agree with you that it should be the default behaviour of the rule. I will make sure to edit and remove the option so that the default behaviour reflects it. Please let me know in the next pull request if I made appropriate changes.

File: tests/lib/rules/jsx-sort-props.js [hunk 1]
```diff
@@ -821,5 +822,231 @@ ruleTester.run('jsx-sort-props', rule, {
...
+        {
+          messageId: 'sortPropsByAlpha',
+          line: 2,
+        },
+        {
+          messageId: 'sortPropsByAlpha',
+          line: 2,
+        },
+      ],
+    } : [],
```
Comment (ljharb) [inline tests/lib/rules/jsx-sort-props.js:1050]:
```suggestion
    } : []
```
now that it's a function call, a trailing function argument comma doesn't work in the older nodes we support

File: lib/rules/jsx-sort-props.js [hunk 2]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
   let groupCount = 0;
+  function addtoSortableAttributeGroups(attribute) {
+    sortableAttributeGroups[groupCount - 1].push(attribute);
+  }
+
   for (let i = 0; i < attributes.length; i++) {
+    const attribute = attributes[i];
+    const nextAttribute = attributes[i + 1];
+    const attributeline = attribute.loc.start.line;
+    let comment = []
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:154]:
```suggestion
    let comment = [];
```

File: lib/rules/jsx-sort-props.js [hunk 3]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
+    try {
+      comment = sourceCode.getCommentsAfter(attribute)
+    } catch (e) {
+    } finally {
+    }
     const lastAttr = attributes[i - 1];
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:160]:
```suggestion
    let comment = [];
    try {
      comment = sourceCode.getCommentsAfter(attribute);
    } catch (e) { /**/  }
    const lastAttr = attributes[i - 1];
```

File: lib/rules/jsx-sort-props.js [hunk 4]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
+        attributeMap.set(attribute, [attribute.range[1], false]);
+        addtoSortableAttributeGroups(attribute);
+      } else {
+        const firstComment = comment[0];
+        const commentline = firstComment.loc.start.line;
+        if (comment.length === 1) {
+          if (attributeline + 1 === commentline && nextAttribute) {
+            attributeMap.set(attribute, [nextAttribute.range[1], true]);
+            addtoSortableAttributeGroups(attribute);
+            i++;
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:184]:
```suggestion
            i += 1;
```

File: lib/rules/jsx-sort-props.js [hunk 5]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
+          } else if (attributeline === commentline) {
+            if (firstComment.type === 'Block') {
+              attributeMap.set(attribute, [nextAttribute.range[1], true]);
+              i++;
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:188]:
```suggestion
              i += 1;
```

File: lib/rules/jsx-sort-props.js [hunk 6]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
+          const commentNextAttribute = sourceCode.getCommentsAfter(nextAttribute);
+          attributeMap.set(attribute, [nextAttribute.range[1], true]);
+          if (
+            commentNextAttribute.length === 1
+            && nextAttribute.loc.start.line === commentNextAttribute[0].loc.start.line
+          ) {
+            attributeMap.set(attribute, [commentNextAttribute[0].range[1], true]);
+          }
+          addtoSortableAttributeGroups(attribute);
+          i++;
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:204]:
```suggestion
          i += 1;
```

**Review by ljharb — COMMENTED** [review]
adding some lint fixes

File: lib/rules/jsx-sort-props.js [hunk 7]
```diff
@@ -118,32 +135,81 @@ function contextCompare(a, b, options) {
...
   for (let i = 0; i < attributes.length; i++) {
+    const attribute = attributes[i];
+    const nextAttribute = attributes[i + 1];
+    const attributeline = attribute.loc.start.line;
+    let comment = []
+    try {
+      comment = sourceCode.getCommentsAfter(attribute)
+    } catch (e) {
+    } finally {
+    }
```
Comment (ljharb) [inline lib/rules/jsx-sort-props.js:159]:
```suggestion
    let comment = [];
    try {
      comment = sourceCode.getCommentsAfter(attribute);
    } catch (e) { /**/ }
```

**Review by ljharb — APPROVED** [review]

**Comment by ROSSROSALES** [comment]
Thank you for reviewing the changes! @ljharb

**Pull request merged**